### PR TITLE
Consolidating `.gitattributes` in one single file

### PR DESCRIPTION
**Pull request by vhvb1989**
fix https://github.com/Azure/azure-dev/issues/324

### CODE REVIEW

**Review by ellismg — CHANGES_REQUESTED** [review]
This change means we no longer include a `.gitattributes` file in any of the generated templates which isn't what we want.

I think we need to retain `templates/common/.gitattributes` as the `.gitattributes` we use across all our templates.  Having some small duplication is fine.

File: .gitattributes [hunk 1]
```diff
@@ -1,2 +1,6 @@
-*.go text eol=lf 
-*.txt text eol=lf
\ No newline at end of file
+**/*.go text eol=lf
+**/*.txt text eol=lf
+**/*.sh text eol=lf
```
Comment (weikanglim) [inline .gitattributes:3]: Does it makes sense to keep one `.gitattributes` file kept separate for the template repositories?

Comment (wbreza) [reply .gitattributes:3]: Agreed - When the templates are generated it should include its own `.gitattributes` file.  We can either use repoman to include it or add any template specific files into the templates.

We should figure this out before moving this forward.

**Review by wbreza — CHANGES_REQUESTED** [review]
Good start but we need to figure out how the generated templates continue to function as expected.

**Comment by vhvb1989** [comment]
> This change means we no longer include a `.gitattributes` file in any of the generated templates which isn't what we want.
> 
> I think we need to retain `templates/common/.gitattributes` as the `.gitattributes` we use across all our templates. Having some small duplication is fine.

@ellismg , @wbreza, @weikanglim  by adding this .gitattributes file to the templates, are we telling customers that they need/must have this file in order to use the template?

The way I understand this file is a way to control working in a team where developers have both Windows and Unix systems, and we want to avoid end-line issues for text files.  

I have the feeling that a file like this should not be injected into a customer's bootstrap project.
It adds complexity for people who doesn't even know about this, to wonder why is this file require for the template...

I think that, within our own team, we want to control the line ending to be consistent in all the text files we add to templates, but do we want to ask customers to use this settings?  I don't see how that is important for the template. 

What are your thoughts?

**Comment by weikanglim** [comment]
> @ellismg , @wbreza, @weikanglim by adding this .gitattributes file to the templates, are we telling customers that they need/must have this file in order to use the template?

I don't think we're telling them one way or the other. The template is a starting point for developers to start their own projects.

We definitely shouldn't be enforcing "non-standard" behavior or be overly controlling when it comes to what goes into the template `.gitattributes` (example: forcing all lines to be `LF` or `CRLF`). But if the `.gitattributes` serve a general purpose, like enforcing line-ending on OS specific scripts (.cmd, .bat and .sh as we currently do), it makes sense to include it. I think along the lines of "what can we do to help the developer accelerate development", and having a `.gitattributes` that help achieve that goal makes sense.

**Comment by vhvb1989** [comment]
> > @ellismg , @wbreza, @weikanglim by adding this .gitattributes file to the templates, are we telling customers that they need/must have this file in order to use the template?
> 
> I don't think we're telling them one way or the other. The template is a starting point for developers to start their own projects.
> 
> We definitely shouldn't be enforcing "non-standard" behavior or be overly controlling when it comes to what goes into the template `.gitattributes` (example: forcing all lines to be `LF` or `CRLF`). But if the `.gitattributes` serve a general purpose, like enforcing line-ending on OS specific scripts (.cmd, .bat and .sh as we currently do), it makes sense to include it. I think along the lines of "what can we do to help the developer accelerate development", and having a `.gitattributes` that help achieve that goal makes sense.

@weikanglim , I am worry about overriding git attibutes from customers.  For example, as a customer, I could have a `/source` folder with all my projects on it. I can add `.gitattributes` to that folder and all my projects would follow the rules from there.
If that's the case, cloning a template would bring its own attributes...

By having .gitattributes for Azure CLI repo, we make sure to ship consistent code into templates.  IMHO, we should let customers to add and create their own coding conventions, like linters, code-formatters, git-hooks, git-attributes, etc.

**Pull request closed without merging**